File: backend/internal/sales/repository/checkout_answers.go
package repository

import (
	"context"
	"database/sql"
	"time"

	"github.com/peter/ticket_pos/backend/internal/catalog"
)

// The checkout half of the Answer's life: the questions a cart is asked, the
// Answers held on the Payment across the Payment Provider redirect, and the copy
// onto the minted Tickets when the sale commits (#311, ADR 0044).
//
// WHY THE SALES REPOSITORY READS `ticket_questions` DIRECTLY. It is the same
// arrangement this file's neighbours already have with `ticket_types`, which the
// sales domain owns nothing of and reads on every checkout: the commit spine
// locks those rows FOR UPDATE, prices from them and decrements their sold_count.
// A seam back into the catalog service for the questions would be a second way
// into a read that has to happen inside this module's own transaction anyway,
// and — for the copy below — inside the very transaction that mints the Tickets.
// What the sales module deliberately does NOT own is the RULES: which answers
// are keepable is catalog.HoldableCheckoutAnswers' finding, and this file only
// moves rows.

// ListCheckoutQuestions returns the Ticket Questions a cart's Ticket Types put
// to a buyer AT CHECKOUT, each with the Options it currently offers.
//
// The query and the fold are catalog.CheckoutQuestionsSQL and
// catalog.ScanAskedQuestions, SHARED WITH THE STOREFRONT'S EVENT PAGE, which
// draws the form these answers come back from. Two copies of those filters would
// eventually differ, and the way that failure shows up is a question the buyer
// was shown whose answer is then silently dropped — or one they were never shown
// that this capture happily accepts.
//
// One read for every Ticket Type in the cart rather than one per type: a cart is
// at most a handful of Ticket Types.
func (r *Repository) ListCheckoutQuestions(ctx context.Context, ticketTypeIDs []string) ([]catalog.AskedQuestion, error) {
	if len(ticketTypeIDs) == 0 {
		return nil, nil
	}
	rows, err := r.db.Pool.QueryContext(ctx, catalog.CheckoutQuestionsSQL, ticketTypeIDs)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return catalog.ScanAskedQuestions(rows)
}

// HoldCheckoutAnswers writes the Answers a buyer gave at checkout onto their
// Payment, keyed by (payment line, index).
//
// It resolves the Payment Line for each answer's Ticket Type ITSELF, from the
// Payment this call names, rather than taking line ids from a caller. That keeps
// the (payment_line, index) key an internal fact: the browser names a Ticket
// Type — which is what it was shown — and the line that Ticket Type became is
// the server's own bookkeeping, aggregated across a cart that may have named the
// same type twice (see BeginCheckout).
//
// AN ANSWER FOR A TICKET TYPE WITH NO PAYMENT LINE IS SKIPPED, not refused. It
// cannot normally happen — the rule that produced these answers already checked
// the cart — but this is the one place on the path where a refusal would have
// somewhere to go, and it must not.
//
// IT RUNS IN ITS OWN TRANSACTION, DELIBERATELY, AND NOT IN CreatePayment'S.
// Writing these beside the payment lines would make a failure here a failure of
// the whole checkout, which is precisely what ADR 0044 forbids. Separated, the
// worst case is a Payment with no held Answers — the buyer keeps their tickets
// and answers later by Answer Link, which is the fallback the whole feature is
// built around. All-or-nothing WITHIN the answers, though: a half-written set is
// a form that silently lost three of its four fields.
func (r *Repository) HoldCheckoutAnswers(ctx context.Context, paymentID string, held []catalog.HeldAnswer, now time.Time) error {
	if len(held) == 0 {
		return nil
	}

	lineIDs, err := r.paymentLineIDsByTicketType(ctx, paymentID)
	if err != nil {
		return err
	}

	tx, err := r.db.Pool.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback() }()

	for _, answer := range held {
		lineID, ok := lineIDs[answer.TicketTypeID]
		if !ok {
			continue
		}
		params := heldAnswerColumns(answer.Value)
		var answerID string
		// The casts sit on the placeholders, exactly as they do in
		// UpsertTicketAnswer: `$5` arrives as a string carrying digits, and
		// telling Postgres it is a NUMERIC is what keeps the value typed all the
		// way in. A driver-side float would round it away.
		//
		// ON CONFLICT DO UPDATE rather than DO NOTHING: a buyer who goes back and
		// re-submits the checkout form for the same Payment means the second
		// answer, and a conflict here must never surface as an error.
		if err := tx.QueryRowContext(ctx, `
			INSERT INTO payment_ticket_answers (
				payment_line_id, ticket_index, ticket_question_id,
				text_value, number_value, date_value, boolean_value, created_at
			)
			VALUES ($1, $2, $3, $4, $5::numeric, $6::date, $7, $8)
			ON CONFLICT (payment_line_id, ticket_index, ticket_question_id) DO UPDATE
			SET text_value = EXCLUDED.text_value,
			    number_value = EXCLUDED.number_value,
			    date_value = EXCLUDED.date_value,
			    boolean_value = EXCLUDED.boolean_value
			RETURNING id
		`, lineID, answer.TicketIndex, answer.TicketQuestionID,
			params.Text, params.Number, params.Date, params.Checked, now,
		).Scan(&answerID); err != nil {
			return err
		}

		// Replaced wholesale, for UpsertTicketAnswer's reason: a choice Answer IS
		// its set of Options, so a re-submission is "this is the set now" and a
		// diff would only add a way to end up somewhere in between.
		if _, err := tx.ExecContext(ctx, `
			DELETE FROM payment_ticket_answer_options WHERE payment_ticket_answer_id = $1
		`, answerID); err != nil {
			return err
		}
		for i, option := range answer.Options {
			if _, err := tx.ExecContext(ctx, `
				INSERT INTO payment_ticket_answer_options (
					payment_ticket_answer_id, ticket_question_option_id,
					option_label_snapshot, sort_order, created_at
				)
				VALUES ($1, $2, $3, $4, $5)
			`, answerID, option.TicketQuestionOptionID, option.LabelSnapshot, i, now); err != nil {
				return err
			}
		}
	}

	return tx.Commit()
}

// paymentLineIDsByTicketType maps a Payment's Ticket Types to its line ids.
//
// One line per Ticket Type is an invariant of CreatePayment, not of the schema:
// BeginCheckout aggregates a cart's lines per Ticket Type before writing them,
// so a cart naming the same type twice is one line. If that ever stopped being
// true this map would silently keep the last line, which is why the aggregation
// is asserted where it happens rather than assumed here.
func (r *Repository) paymentLineIDsByTicketType(ctx context.Context, paymentID string) (map[string]string, error) {
	rows, err := r.db.Pool.QueryContext(ctx, `
		SELECT id, ticket_type_id FROM payment_lines WHERE payment_id = $1
	`, paymentID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	byType := map[string]string{}
	for rows.Next() {
		var id, ticketTypeID string
		if err := rows.Scan(&id, &ticketTypeID); err != nil {
			return nil, err
		}
		byType[ticketTypeID] = id
	}
	return byType, rows.Err()
}

// LineAnswer is one held Answer on its way from a Payment onto a Ticket: which
// unit of the line it is about, which question it replies to, and the reply in
// the columns it is already stored in.
//
// IT CARRIES NO KIND AND DOES NO PARSING, and that is the point of mirroring the
// typed columns on both tables (migration 074). By the time an Answer is here it
// has been through catalog.ParseAnswer once, at checkout, and the write below is
// a COPY. A second parse at commit would be a second chance to disagree with the
// first — after the Payment Provider has taken the money, where a disagreement
// has nowhere to go.
type LineAnswer struct {
	// TicketIndex is which of the line's units this Answer is about, 1..quantity.
	// It becomes `tickets.ordinal` unchanged; see writeTicketAnswers.
	TicketIndex      int
	TicketQuestionID string
	Text             sql.NullString
	// Number and Date are strings out of NUMERIC and DATE columns, so the digits
	// travel exactly as Postgres holds them — `3.50` keeps its trailing zero, and
	// a calendar date never acquires a midnight or a zone.
	Number  sql.NullString
	Date    sql.NullString
	Checked sql.NullBool
	Options []LineAnswerOption
}

// LineAnswerOption is one chosen Option travelling with its Answer: the Option's
// identity, and the SNAPSHOT taken when the buyer read it at checkout.
//
// The snapshot is carried rather than retaken. Between checkout and commit sits
// a Payment Provider redirect, which is the one gap in this product long enough
// for an Organization to rename an Option; re-reading the label here would put
// words on the Ticket that nobody was ever shown (migration 074).
type LineAnswerOption struct {
	TicketQuestionOptionID string
	LabelSnapshot          string
}

// ListHeldAnswersByPaymentLine returns the Answers a Payment is holding, grouped
// by the Payment Line they were given against.
//
// Read inside the commit transaction, so the copy onto the Tickets and the
// minting of those Tickets cannot come apart.
func listHeldAnswersByPaymentLine(ctx context.Context, tx *sql.Tx, paymentID string) (map[string][]LineAnswer, error) {
	rows, err := tx.QueryContext(ctx, `
		SELECT a.payment_line_id, a.id, a.ticket_index, a.ticket_question_id,
		       a.text_value, a.number_value::text, a.date_value::text, a.boolean_value
		FROM payment_ticket_answers a
		JOIN payment_lines l ON l.id = a.payment_line_id
		WHERE l.payment_id = $1
		ORDER BY a.payment_line_id, a.ticket_index, a.created_at
	`, paymentID)
	if err != nil {
		return nil, err
	}

	byLine := map[string][]LineAnswer{}
	// Where each Answer ended up, so its Options can be attached in the second
	// pass without a second lookup per row.
	type position struct {
		lineID string
		index  int
	}
	positions := map[string]position{}
	for rows.Next() {
		var lineID, answerID string
		var answer LineAnswer
		if err := rows.Scan(&lineID, &answerID, &answer.TicketIndex, &answer.TicketQuestionID,
			&answer.Text, &answer.Number, &answer.Date, &answer.Checked); err != nil {
			rows.Close()
			return nil, err
		}
		byLine[lineID] = append(byLine[lineID], answer)
		positions[answerID] = position{lineID: lineID, index: len(byLine[lineID]) - 1}
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	if len(positions) == 0 {
		return byLine, nil
	}

	// TWO QUERIES FOR ANY NUMBER OF ANSWERS, not two per Answer — the shape
	// ListTicketAnswers uses for the same reason. A cart of forty tickets with
	// six choice questions each is one read of the Answers and one of their
	// Options.
	optionRows, err := tx.QueryContext(ctx, `
		SELECT o.payment_ticket_answer_id, o.ticket_question_option_id, o.option_label_snapshot
		FROM payment_ticket_answer_options o
		JOIN payment_ticket_answers a ON a.id = o.payment_ticket_answer_id
		JOIN payment_lines l ON l.id = a.payment_line_id
		WHERE l.payment_id = $1
		ORDER BY o.payment_ticket_answer_id, o.sort_order, o.created_at
	`, paymentID)
	if err != nil {
		return nil, err
	}
	defer optionRows.Close()
	for optionRows.Next() {
		var answerID string
		var option LineAnswerOption
		if err := optionRows.Scan(&answerID, &option.TicketQuestionOptionID, &option.LabelSnapshot); err != nil {
			return nil, err
		}
		at, ok := positions[answerID]
		if !ok {
			continue
		}
		byLine[at.lineID][at.index].Options = append(byLine[at.lineID][at.index].Options, option)
	}
	return byLine, optionRows.Err()
}

// writeTicketAnswers copies a line's held Answers onto the Tickets that line
// just minted, in the caller's transaction.
//
// `ticketIDs` maps ordinal to Ticket id, straight out of mintTickets. THE WHOLE
// OF "IN ORDER" IS THIS LOOKUP: index n on the Payment becomes ordinal n on the
// Ticket, and neither number is ever recomputed from a row position. That is
// what `tickets.ordinal` was given a column for (migration 070) — without it the
// order would be whatever the INSERT happened to return.
//
// AN ANSWER WHOSE INDEX NAMES NO TICKET IS SKIPPED. The capture already bounded
// the index by the line's quantity, so this can only be reached by a Payment
// whose line was somehow written with a smaller quantity than the answers it
// holds. Skipping is the only safe way to be wrong: failing here would fail the
// commit of a sale the Payment Provider has already been paid for, which is the
// PAYMENT_APPROVED_WITHOUT_SALE incident, manufactured over an answer.
//
// It runs in the SAME transaction as the minting, and nothing between the two
// may fail without both rolling back: a Payment that fails or expires produces
// no Tickets, and therefore no Answers on any Ticket.
func writeTicketAnswers(ctx context.Context, tx *sql.Tx, ticketIDs map[int]string, answers []LineAnswer, now time.Time) error {
	for _, answer := range answers {
		ticketID, ok := ticketIDs[answer.TicketIndex]
		if !ok {
			continue
		}
		var answerID string
		// A plain INSERT and not an upsert: these Tickets were minted moments
		// ago in this same transaction and can have no Answers yet. An upsert
		// here would quietly paper over a double-mint, which the UNIQUE on
		// `tickets` exists to make loud.
		if err := tx.QueryRowContext(ctx, `
			INSERT INTO ticket_answers (
				ticket_id, ticket_question_id,
				text_value, number_value, date_value, boolean_value,
				created_at, updated_at
			)
			VALUES ($1, $2, $3, $4::numeric, $5::date, $6, $7, $7)
			RETURNING id
		`, ticketID, answer.TicketQuestionID,
			answer.Text, answer.Number, answer.Date, answer.Checked, now,
		).Scan(&answerID); err != nil {
			return err
		}
		for i, option := range answer.Options {
			// The snapshot travels verbatim from the Payment. The Option's
			// CURRENT label is what every list and the export header show, and
			// it comes from the join; this is what the buyer read.
			if _, err := tx.ExecContext(ctx, `
				INSERT INTO ticket_answer_options (
					ticket_answer_id, ticket_question_option_id,
					option_label_snapshot, sort_order, created_at
				)
				VALUES ($1, $2, $3, $4, $5)
			`, answerID, option.TicketQuestionOptionID, option.LabelSnapshot, i, now); err != nil {
				return err
			}
		}
	}
	return nil
}

// heldAnswerColumns turns one validated AnswerValue into the four nullable
// columns it is stored in, setting exactly the one its kind takes.
//
// It is the checkout twin of the catalog service's upsertParams, and the two
// agree on the one thing that matters: a checkbox Answer is written Valid even
// when it is FALSE, because false is an answer. A NULL there would be
// indistinguishable from a question nobody replied to, which is an Outstanding
// Answer and a different fact.
func heldAnswerColumns(value catalog.AnswerValue) struct {
	Text    sql.NullString
	Number  sql.NullString
	Date    sql.NullString
	Checked sql.NullBool
} {
	var params struct {
		Text    sql.NullString
		Number  sql.NullString
		Date    sql.NullString
		Checked sql.NullBool
	}
	switch value.Kind {
	case catalog.TicketQuestionKindShortText, catalog.TicketQuestionKindLongText:
		params.Text = sql.NullString{String: value.Text, Valid: true}
	case catalog.TicketQuestionKindNumber:
		params.Number = sql.NullString{String: value.Number, Valid: true}
	case catalog.TicketQuestionKindDate:
		params.Date = sql.NullString{String: value.Date, Valid: true}
	case catalog.TicketQuestionKindCheckbox:
		params.Checked = sql.NullBool{Bool: value.Checked, Valid: true}
	}
	return params
}

// PurgeAbandonedCheckoutAnswers deletes the Answers held on Payments that never
// reached 'approved' and were begun at or before `cutoff` (#316, ADR 0044). It
// reports how many Answers went and how many Payments they came off.
//
// THE PREDICATE IS `status IS DISTINCT FROM 'approved'` AND NOTHING ELSE, and
// the reason it is not `status = 'expired'` is the whole ticket. Two facts about
// this codebase make the status useless as a "this is over" signal, and each one
// breaks a different half of the naive query:
//
//   - 'expired' IS NOT TERMINAL. It is opportunistic bookkeeping written by
//     whichever begin-checkout happens to pass the same Event
//     (ExpireStalePayments), and ApprovePaymentAndCommitSale accepts an expired
//     Payment exactly like a pending one — a provider that confirms late still
//     commits the sale, flipping expired → approved. Purging on 'expired' would
//     delete the Answers of a sale that then commits, silently: the copy at
//     commit finds nothing, the buyer keeps their Tickets, and the Organization
//     is simply told nobody answered.
//
//   - 'pending' IS OFTEN FOREVER. Nothing sweeps Payments on its own schedule,
//     so an Event with no further traffic keeps its abandoned Payments 'pending'
//     for good. A purge that waited for 'expired' would never touch them, and
//     the quietest Events — the ones with one abandoned checkout and nobody
//     looking — would be exactly the ones that kept the health data.
//
// So the status is read only to EXCLUDE the approved, and it is AGE that decides
// the rest. `IS DISTINCT FROM` rather than `<>` so a status that ever became
// nullable would purge rather than silently exempt: this is a deletion, and the
// safe direction for it to be wrong is loudly, not by quietly retaining.
//
// AGE IS THE PAYMENT'S, NOT THE ANSWER'S. `payment_ticket_answers.created_at`
// moves when a buyer goes back and re-submits the checkout form (HoldCheckoutAnswers
// upserts), so keying on it would restart the clock for a buyer who edited their
// answer and then abandoned — the retention window would be about typing rather
// than about the attempt to buy. The rule is "30 days after the Payment".
//
// The chosen Options go with their Answers by the ON DELETE CASCADE on
// `payment_ticket_answer_options` (migration 074). That cascade is why this is
// one statement and not two, and it is worth knowing that deleting the Answers
// without it would leave the option labels — the words somebody picked — behind.
//
// IT DELETES ONLY FROM `payment_ticket_answers`. The Payment and its
// `payment_lines` are untouched and are kept forever: the platform is entitled
// to remember an attempt to transact, and what it may not keep is the reply to a
// question about a Ticket that will never exist.
//
// IDEMPOTENT BY CONSTRUCTION, because it is a DELETE of rows matched by a
// predicate rather than a state machine: a second run finds the rows gone and
// reports zeros, and two runs racing each other delete disjoint sets. It takes
// no lock and claims nothing — there is no queue here and no per-row bookkeeping
// to leave behind, which is what distinguishes it from the Reversal Reconciler's
// drain.
func (r *Repository) PurgeAbandonedCheckoutAnswers(ctx context.Context, cutoff time.Time) (answers, payments int64, err error) {
	// One statement, in two CTEs. `doomed` names the rows and the Payments they
	// belong to BEFORE the delete, which is the only moment the Payment can still
	// be counted — a DELETE ... RETURNING gives back the answer rows, and by then
	// there is nothing left to join to the line that says whose they were.
	err = r.db.Pool.QueryRowContext(ctx, `
		WITH doomed AS (
			SELECT a.id, l.payment_id
			FROM payment_ticket_answers a
			JOIN payment_lines l ON l.id = a.payment_line_id
			JOIN payments p ON p.id = l.payment_id
			WHERE p.status IS DISTINCT FROM 'approved'
			  AND p.created_at <= $1
		),
		purged AS (
			DELETE FROM payment_ticket_answers
			WHERE id IN (SELECT id FROM doomed)
			RETURNING id
		)
		SELECT (SELECT COUNT(*) FROM purged), (SELECT COUNT(DISTINCT payment_id) FROM doomed)
	`, cutoff).Scan(&answers, &payments)
	if err != nil {
		return 0, 0, err
	}
	return answers, payments, nil
}

// CountHeldCheckoutAnswers is how many Answers are riding Payments right now,
// across the platform.
//
// Reported by every purge run for the reason the Reversal Reconciler reports its
// in-flight backlog: a run that says "0 purged" is indistinguishable from a run
// against a table that was never written to, and an operator watching a feature
// that ships dark (ADR 0045) needs to tell "nothing was due" from "nothing is
// there". Two curls a day apart say whether the checkout is capturing at all.
func (r *Repository) CountHeldCheckoutAnswers(ctx context.Context) (int64, error) {
	var n int64
	err := r.db.Pool.QueryRowContext(ctx, `SELECT COUNT(*) FROM payment_ticket_answers`).Scan(&n)
	return n, err
}

File: apps/storefront/lib/checkout-answers.ts
/**
 * The checkout's answer section, as pure data (#311, ADR 0044).
 *
 * A Ticket Question asks about the person who will HOLD a ticket, and the buyer
 * of four tickets is not assumed to know four people's answers. So the section
 * is one set of questions PER TICKET, entirely skippable, and nothing in it can
 * stop a purchase: no validation lives here, no required check, no disabled
 * button. A question the buyer leaves alone becomes an Outstanding Answer that
 * the Ticket's holder can fill in later through its Answer Link.
 *
 * Framework-free, like checkout.ts beside it, so the decisions are unit tested
 * and the component owns only the drawing.
 */

/** One selectable value of a choice question: its stable id, and its words. */
export type CheckoutQuestionOption = {
  /**
   * What an answer NAMES. Never the label — a label cannot survive a rename,
   * which is the whole reason an Option has an id.
   */
  id: string;
  label: string;
};

/** One Ticket Question as the Event page received it. */
export type CheckoutQuestion = {
  id: string;
  /**
   * The Organization's own words, read as coined in every Locale exactly as a
   * Custom Tag is (ADR 0027). Only the chrome around it is translated.
   */
  label: string;
  /** Which of the seven fields to draw. */
  kind:
    | "short_text"
    | "long_text"
    | "single_choice"
    | "multi_choice"
    | "number"
    | "date"
    | "checkbox";
  /**
   * MARKS THE FIELD AND GATES NOTHING. Required's only effect anywhere is
   * producing an Outstanding Answer the Organization can chase; a Storefront
   * that turned it into a blocked pay button would be reversing ADR 0044, whose
   * premise is that the buyer often does not know the answer.
   */
  required: boolean;
  /** Empty for the five kinds that are not answered by choosing. */
  options: CheckoutQuestionOption[];
};

/** As much of a Ticket Type as the answer section needs. */
export type AnsweredTicketType = {
  id: string;
  name: string;
  /**
   * The effective buyer price per ticket, exactly as the Event page received it:
   * server-computed, fee included where the Event passes it on, and ALREADY the
   * Promotional Price where a Promotion is live. It is here because the buyer's
   * own Ticket is the dearest one in the cart (ADR 0074), and the price that
   * decides that is the price as sold — this app never recomputes either.
   */
  price_cents: number;
  /**
   * Absent on a deployment where the feature flag is closed, which is how it
   * ships — the API omits the key entirely, so there is nothing to draw and no
   * second flag on this side to disagree with it (ADR 0045).
   */
  ticket_questions?: CheckoutQuestion[];
};

/** One ticket in the cart, and what it is being asked. */
export type AnswerSlot = {
  ticketTypeId: string;
  ticketTypeName: string;
  /**
   * Which of that Ticket Type's tickets this is, ONE-BASED, counted across the
   * whole cart's holding of it. It is sent as `ticket_index` and becomes the
   * minted Ticket's `ordinal`, which is what makes "the answers land in order"
   * mean anything — so it starts at 1 here because it starts at 1 there.
   */
  index: number;
  questions: CheckoutQuestion[];
};

/** One reply, in the one slot its question's kind takes. */
export type AnswerValue = {
  text?: string;
  /**
   * As the digits were TYPED. A string and not a number: this lands in a NUMERIC
   * column, and round-tripping it through JavaScript's float is how `0.1`
   * becomes `0.100000001` and `3.50` loses the zero it meant.
   */
  number?: string;
  date?: string;
  checked?: boolean;
  option_ids?: string[];
};

/** Every reply on the form, keyed by answerKey. */
export type AnswerValues = Record<string, AnswerValue>;

/** One entry of the begin-checkout body's answer section. */
export type CheckoutAnswerBody = {
  ticket_type_id: string;
  ticket_index: number;
  ticket_question_id: string;
} & AnswerValue;

/**
 * answerKey identifies one ticket's reply to one question.
 *
 * All three parts are needed and none is redundant: two tickets of one Ticket
 * Type differ only in the index, and two questions on one ticket differ only in
 * the question. A key missing either would silently merge two fields into one.
 */
export function answerKey(ticketTypeId: string, index: number, questionId: string): string {
  return `${ticketTypeId}:${index}:${questionId}`;
}

/**
 * answerSlots expands the cart into one set of questions per ticket.
 *
 * THIS IS THE FEATURE'S SHAPE IN ONE FUNCTION. A cart line of three becomes
 * three sets, because an Answer belongs to a Ticket and a buyer of three may
 * have three different answers (ADR 0043). A Ticket Type asking nothing, and one
 * nobody is buying, contribute no sets at all — so a cart of ordinary tickets
 * has no answer section and the dialog is the one that existed before this
 * feature.
 *
 * The order is the page's own order of Ticket Types, then the ticket's number,
 * so the form reads the way the cart does.
 */
export function answerSlots(
  ticketTypes: AnsweredTicketType[],
  quantities: Record<string, number>,
): AnswerSlot[] {
  const slots: AnswerSlot[] = [];
  for (const ticketType of ticketTypes) {
    const questions = ticketType.ticket_questions ?? [];
    if (questions.length === 0) continue;
    const quantity = quantities[ticketType.id] ?? 0;
    for (let index = 1; index <= quantity; index++) {
      slots.push({
        ticketTypeId: ticketType.id,
        ticketTypeName: ticketType.name,
        index,
        questions,
      });
    }
  }
  return slots;
}

/**
 * ownTicketSlot is the one ticket checkout asks about: the buyer's own, which
 * the sale will hand them as a Self-held Ticket (ADR 0048). It is the FIRST
 * ticket of the DEAREST Ticket Type the cart holds, ties broken by the catalog's
 * order (ADR 0074) — the same choice the commit spine makes, so what the dialog
 * calls "your ticket" is the Ticket the buyer will in fact hold.
 *
 * THE TWO RULES MUST STAY ONE SENTENCE. A buyer seated by the backend on one
 * Ticket and asked another Ticket's questions by this page answers for somebody
 * who is not them, and leaves their own Ticket owing — which is the bug ADR 0074
 * was written about, in its other half.
 *
 * The price compared is `price_cents`: what this buyer pays, Promotional Price
 * and all, never a List Price. The catalog arrives in its own order, so keeping
 * the first of the equals IS the tie-break.
 *
 * THE OTHER TICKETS ARE NOT ASKED ABOUT AT ALL, and are not mentioned. A buyer
 * of four is not assumed to know four people's sizes; those Tickets are for
 * their own Holders to answer, after the purchase, from the sale page's links.
 *
 * Null when the sale will hand the buyer nothing — the assignment feature is
 * closed, the cart is empty — or when the buyer's own Ticket Type asks nothing:
 * then there is no section, and the dialog is the one that existed before
 * Ticket Questions did.
 */
export function ownTicketSlot(
  ticketTypes: AnsweredTicketType[],
  quantities: Record<string, number>,
  buyerHoldsFirstTicket: boolean,
): AnswerSlot | null {
  if (!buyerHoldsFirstTicket) return null;
  let own: AnsweredTicketType | undefined;
  for (const ticketType of ticketTypes) {
    if ((quantities[ticketType.id] ?? 0) <= 0) continue;
    if (own === undefined || ticketType.price_cents > own.price_cents) own = ticketType;
  }
  if (own === undefined) return null;
  const questions = own.ticket_questions ?? [];
  if (questions.length === 0) return null;
  return { ticketTypeId: own.id, ticketTypeName: own.name, index: 1, questions };
}

/** The Upgrade Prompt, as the dialog needs it: whether to draw it, and about what. */
export type UpgradePrompt = {
  /**
   * The cart line the buyer gives up by electing, or NULL when the free Ticket
   * in play is on an earlier Sale and the cart holds none.
   *
   * It is the only thing the two mechanisms differ by that this app may know,
   * and it exists here for one reason: the prompt has to say something true, and
   * "the free ticket in your basket" and "the free ticket you already have" are
   * not the same sentence. The MECHANISM behind them — a line never bought
   * versus a Sale reversed inside the paid commit — is never mentioned to the
   * buyer, by ADR 0074, and this field must not become a way to mention it.
   *
   * NOTHING IS DROPPED FROM `lines` ON ACCOUNT OF IT. The cart is sent exactly
   * as the buyer built it and the commit drops the free line itself, so this
   * names what WILL happen rather than something this app does. A Storefront
   * that pre-emptively removed the line would be a second implementation of the
   * ambiguity rule, and would sell one Ticket fewer the moment the two disagreed.
   */
  surrendered: { ticketTypeId: string; ticketTypeName: string } | null;
};

/**
 * upgradePrompt decides whether the checkout offers an Upgrade, and about which
 * free Ticket (ADR 0074, #652).
 *
 * THIS RESTATES A BACKEND RULE, AND IT IS THE ONE PLACE IN THE FEATURE WHERE
 * DRIFT CAN ENTER. The Go counterpart is
 * `UpgradeEligibility.OffersUpgrade(freeTicketsInBasket)` in
 * `backend/internal/sales/repository/self_held.go` —
 * `surrenderable_free_tickets + free Tickets in the basket == 1` — and the
 * duplication is unavoidable rather than sloppy: the basket is client-side and
 * never reaches the Event-page request that publishes the count, so neither side
 * can do this arithmetic alone. Nothing but review keeps the two sentences the
 * same. It is written ONCE, here, so that there is only ever one of them to
 * check, and the component above it decides nothing.
 *
 * THE COUNT IS NULLABLE AND NULL IS NOT ZERO. Null is the anonymous read — "we
 * do not know who is asking" — and yields no prompt: there is no buyer whose
 * earlier Sales were counted, so there is no offer to make. Zero is a real
 * answer given to somebody we can identify, and a zero with one free Ticket in
 * the cart is a genuine offer. Reading null as 0 would show the prompt to
 * anonymous readers, and it is the only mistake here that produces a plausible-
 * looking page rather than a visible one.
 *
 * NOTHING IT RETURNS CAN FAIL A PURCHASE. A prompt drawn where the commit will
 * not honour it costs nothing — an election the backend did not offer is IGNORED
 * and never refused — and a prompt withheld where the commit would have honoured
 * it costs a buyer nothing but the offer. That asymmetry is why the rule may
 * live in two places at all.
 *
 * Three things must hold, and each is a case ADR 0074 argued:
 *
 *   - THE CART HOLDS SOMETHING PAID. An Upgrade is free to paid, so without a
 *     paid Ticket to move onto there is nothing to elect. The backend's
 *     OffersUpgrade deliberately does not carry this precondition, leaving it to
 *     whichever surface does the offering; on this side, that is this function.
 *   - EXACTLY ONE FREE TICKET IS IN PLAY, counting the cart and the buyer's
 *     earlier Sales together. Two is ambiguous from either side, and the answer
 *     to ambiguity is no prompt rather than a guess.
 *   - FREE MEANS `price_cents === 0`, which is the price AS SOLD: the API has
 *     already applied any Promotional Price, and this app never recomputes a
 *     price. A Ticket Type given away today is free in this basket, which is the
 *     honest reading of "cost nothing" and the one `unit_price_cents` takes.
 *
 * Counted over QUANTITY throughout: a line of two free Tickets is two free
 * Tickets. Since exactly one free Ticket in the cart is therefore exactly one
 * free LINE, naming the line and naming the Ticket are the same act here.
 */
export function upgradePrompt(
  ticketTypes: AnsweredTicketType[],
  quantities: Record<string, number>,
  surrenderableFreeTickets: number | null,
): UpgradePrompt | null {
  if (surrenderableFreeTickets === null) return null;

  let freeInBasket = 0;
  let paidInBasket = 0;
  let freeLine: AnsweredTicketType | undefined;
  for (const ticketType of ticketTypes) {
    const quantity = quantities[ticketType.id] ?? 0;
    if (quantity <= 0) continue;
    if (ticketType.price_cents === 0) {
      freeInBasket += quantity;
      freeLine ??= ticketType;
      continue;
    }
    paidInBasket += quantity;
  }

  if (paidInBasket === 0) return null;
  if (surrenderableFreeTickets + freeInBasket !== 1) return null;

  return {
    surrendered:
      freeLine === undefined
        ? null
        : { ticketTypeId: freeLine.id, ticketTypeName: freeLine.name },
  };
}

/**
 * hasCheckoutQuestions reports whether anything in the cart asks anything —
 * which is what decides whether the dialog draws an answer section's heading at
 * all. It is `answerSlots(...).length > 0` said cheaply and named for what the
 * caller means.
 */
export function hasCheckoutQuestions(
  ticketTypes: AnsweredTicketType[],
  quantities: Record<string, number>,
): boolean {
  return answerSlots(ticketTypes, quantities).length > 0;
}

/**
 * checkoutAnswerBodies turns the form's state into the begin-checkout body's
 * `answers` array, sending only what the buyer actually filled in.
 *
 * IT IS DRIVEN BY THE SLOTS AND NOT BY THE VALUES, deliberately: the slots are
 * the truth about what is being bought, so a reply left behind by a ticket the
 * buyer then removed from the cart simply has no slot and does not travel. The
 * alternative — iterating the values — would send answers for tickets that no
 * longer exist, which the API would drop, having first written them down as a
 * thing somebody typed.
 *
 * A BLANK IS NOT AN ANSWER. An emptied text field and an empty set of chosen
 * Options are both "not said", and the way to say that is to send nothing: a
 * stored blank is a row no later reader could tell from a real reply. A
 * CHECKBOX IS THE EXCEPTION AND THE INTERESTING ONE — `false` is somebody who
 * read the question and said no, so it travels, while a checkbox nobody touched
 * has no value at all and does not.
 *
 * NOTHING HERE VALIDATES. Whether a reply fits its question's kind is the API's
 * finding, and its answer to "no" is to drop the reply rather than refuse the
 * purchase (ADR 0044). A second copy of those rules on this side would be a
 * second place for them to be enforced differently — and this is the side where
 * a refusal is easy to write by accident.
 */
export function checkoutAnswerBodies(
  slots: AnswerSlot[],
  values: AnswerValues,
): CheckoutAnswerBody[] {
  const bodies: CheckoutAnswerBody[] = [];
  for (const slot of slots) {
    for (const question of slot.questions) {
      const value = values[answerKey(slot.ticketTypeId, slot.index, question.id)];
      const stated = statedReply(value);
      if (!stated) continue;
      bodies.push({
        ticket_type_id: slot.ticketTypeId,
        ticket_index: slot.index,
        ticket_question_id: question.id,
        ...stated,
      });
    }
  }
  return bodies;
}

/**
 * statedReply narrows one form value to the reply it actually states, or null
 * when the buyer said nothing.
 *
 * Exactly one slot comes back filled, because exactly one is what the API takes:
 * a body naming two is a caller believing two things about one question, and is
 * refused there rather than coerced. The order below is the order the kinds are
 * checked in and no value can legitimately fill two.
 */
function statedReply(value: AnswerValue | undefined): AnswerValue | null {
  if (!value) return null;
  if (typeof value.checked === "boolean") return { checked: value.checked };
  if (value.option_ids && value.option_ids.length > 0) return { option_ids: value.option_ids };
  if (value.text !== undefined && value.text.trim() !== "") return { text: value.text };
  if (value.number !== undefined && value.number.trim() !== "") return { number: value.number };
  if (value.date !== undefined && value.date.trim() !== "") return { date: value.date };
  return null;
}
